url882f91ec97_WenboCao_Microsoft_Drone_tgz-pJ8-org_ejml_alg_dense_decomposition_qr_StabilityQRDecompositionJ8/src/org/ejml/alg/block/BlockTriangularSolver.java

```diff
@@ -0,0 +1,516 @@
+            Binner.row0 = B.row0 + i;   Binner.row1 = Binner.row0 + widthT;
+
+            // solve the top row block
+            // B(i,:) = T(i,i)^-1 Y(i,:)
+            solveBlock(blockLength,false, Linner,Binner,transL,false);
+
+            boolean updateY;
+            if( transL ) {
+                updateY = Linner.row0 > 0;
+            } else {
+                updateY = Linner.row1 < L.row1;
+            }
+            if( updateY ) {
+                // Y[i,:] = Y[i,:] - sum j=1:i-1 { T[i,j] B[j,i] }
+                // where i is the next block down
+                // The summation is a block inner product
+                if( transL ) {
+                    Linner.col1 = Linner.col0;
+                    Linner.col0 = Linner.col1 - blockLength;
+                    Linner.row1 = L.row1;
+                    //Tinner.col1 = Tinner.col1;
+
+//                    Binner.row0 = Binner.row0;
+                    Binner.row1 = B.row1;
+
+                    Y.row0 = Binner.row0-blockLength;
+                    Y.row1 = Binner.row0;
+                } else {
+                    Linner.row0 = Linner.row1;
+                    Linner.row1 = Math.min(Linner.row0+blockLength, L.row1);
+                    Linner.col0 = L.col0;
+                    //Tinner.col1 = Tinner.col1;
+
+                    Binner.row0 = B.row0;
+                    //Binner.row1 = Binner.row1;
+
+                    Y.row0 = Binner.row1;
+                    Y.row1 = Math.min(Y.row0+blockLength,B.row1);
+                }
+
+                // step through each block column
+                for( int k = B.col0; k < B.col1; k += blockLength ) {
+
+                    Binner.col0 = k;
+                    Binner.col1 = Math.min(k+blockLength,B.col1);
+
+                    Y.col0 = Binner.col0;
+                    Y.col1 = Binner.col1;
+
+                    if( transL ) {
+                        // Y = Y - T^T * B
+                        BlockMultiplication.multMinusTransA(blockLength, Linner,Binner,Y);
+                    } else {
+
+                        // Y = Y - T * B
+                        BlockMultiplication.multMinus(blockLength, Linner,Binner,Y);
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * <p>
+     * Solves upper triangular systems:<br>
+     * <br>
+     * B = R<sup>-1</sup> B<br>
+     * <br>
+     * </p>
+     *
+     * <p>Only the first B.numRows rows in R will be processed.  Lower triangular elements are ignored.<p>
+     *
+     * <p> Reverse or forward substitution is used depending upon L being transposed or not. </p>
+     *
+     * @param blockLength
+     * @param R Upper triangular with dimensions m by m.  Not modified.
+     * @param B A matrix with dimensions m by n.  Solution is written into here. Modified.
+     * @param transR Is the triangular matrix transposed?
+     */
+    public static void solveR( final int blockLength ,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 15. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.